client/widmgr/widmgr.go

```diff
@@ -4,80 +4,367 @@
-			Region:        m.region,
-			MinQueryIndex: minIndex - 1,
-			AllowStale:    true,
-			AuthToken:     m.nodeSecret,
-		},
+// If an error is returned the identities could not be fetched and the renewal
+// goroutine was not started.
+func (m *WIDMgr) Run() error {
+	if len(m.widSpecs) == 0 {
+		m.logger.Debug("no workload identities to retrieve or renew")
+		return nil
 	}
-	reply := structs.AllocIdentitiesResponse{}
-	if err := m.rpc.RPC("Alloc.SignIdentities", &args, &reply); err != nil {
-		return nil, err
+
+	m.logger.Debug("retrieving and renewing workload identities", "num_identities", len(m.widSpecs))
+
+	if err := m.getIdentities(); err != nil {
+		return fmt.Errorf("failed to fetch signed identities: %w", err)
 	}
 
-	if n := len(reply.Rejections); n == 1 {
-		return nil, fmt.Errorf("%d/%d signing request was rejected", n, len(req))
-	} else if n > 1 {
-		return nil, fmt.Errorf("%d/%d signing requests were rejected", n, len(req))
+	go m.renew()
+
+	return nil
+}
+
+// Get retrieves the latest signed identity or returns an error. It must be
+// called after Run and does not block.
+//
+// For retrieving tokens which might be renewed callers should use Watch
+// instead to avoid missing new tokens retrieved by Run between Get and Watch
+// calls.
+func (m *WIDMgr) Get(id cstructs.TaskIdentity) (*structs.SignedWorkloadIdentity, error) {
+	token := m.get(id)
+	if token == nil {
+		// This is an error as every identity should have a token by the time Get
+		// is called.
+		return nil, fmt.Errorf("unable to find token for task %q and identity %q", id.TaskName, id.IdentityName)
 	}
 
-	if len(reply.SignedIdentities) == 0 {
-		return nil, fmt.Errorf("empty signed identity response")
+	return token, nil
+}
+
+func (m *WIDMgr) get(id cstructs.TaskIdentity) *structs.SignedWorkloadIdentity {
+	m.lastTokenLock.Lock()
+	defer m.lastTokenLock.Unlock()
+
+	return m.lastToken[id]
+}
+
+// Watch sends new signed identities until it is closed due to shutdown. Must
+// be called after Run.
+//
+// The caller must call the returned func to stop watching.
+func (m *WIDMgr) Watch(id cstructs.TaskIdentity) (<-chan *structs.SignedWorkloadIdentity, func()) {
+	m.watchersLock.Lock()
+	defer m.watchersLock.Unlock()
+
+	// If Shutdown has been called return a closed chan
+	if m.stopCtx.Err() != nil {
+		c := make(chan *structs.SignedWorkloadIdentity)
+		close(c)
+		return c, func() {}
+	}
+
+	//TODO I think we only want 1 watcher per id, so let's just close on double watches?
+	if existing, ok := m.watchers[id]; ok {
+		close(existing)
+	}
+
+	// Buffer of 1 so sends don't block on receives
+	c := make(chan *structs.SignedWorkloadIdentity, 1)
+	m.watchers[id] = c
+
+	// Create a cancel func for watchers to deregister when they exit.
+	cancel := func() {
+		m.watchersLock.Lock()
+		defer m.watchersLock.Unlock()
+
+		delete(m.watchers, id)
+
+		//TODO should we close(c) here? seems sketchy... caller shouldn't try to
+		//recv on c after calling cancel
```

Comment:
We might want to close here but then we'll need to change the `Shutdown` method so we don't close there. Otherwise during shutdown all the `identity_hook`s in the task runner will call their own `Stop` which stops their watch channels. That'd result in a double-close during shutdown (and panic). We probably wouldn't be able to safely do the "so let's just close on double watches" you've done above either.

It's easier to reason about correctness if we make sure that the close can only happen in exactly one place (and one goroutine) per channel. But we have two ways in which we want to stop watching: (1) the task runner has exited because the task has stopped, and (2) the agent itself is shutting down. The current design you've got here handles (2) just fine and then for (1) we delete the channel from the list of watchers and the taskrunner is responsible for making sure it's not trying to poll on it anymore so the channel can be GC'd.

Reply:
Ok I think I get it. I'll remove the comment.